wiring/src/spark_wiring_logging.cpp

```diff
@@ -0,0 +1,246 @@
+    const char* result = strchr(s, c);
+    return result ? result : s + strlen(s);
+}
+
+} // namespace
+
+// spark::LogHandler
+struct spark::LogHandler::FilterData {
+    const char *name; // Category name
+    size_t size; // Name length
+    int level; // Logging level (-1 if not specified for this category)
+    std::vector<FilterData> filters; // Subcategories
+
+    FilterData(const char *name, size_t size) :
+            name(name),
+            size(size),
+            level(-1) {
+    }
+};
+
+spark::LogHandler::LogHandler(Stream *stream, LogLevel level, const Filters &filters) :
+        stream_(stream),
+        level_(level) {
+    for (auto it = filters.begin(); it != filters.end(); ++it) {
+        const char* const category = it->first;
+        const LogLevel level = it->second;
+        std::vector<FilterData> *filters = &filters_; // Root categories
+        size_t pos = 0;
+        for (size_t i = 0;; ++i) {
+            if (category[i] && category[i] != '.') { // Category separator
+                continue;
+            }
+            const size_t size = i - pos;
+            if (!size) {
+                break; // Invalid category name
+            }
+            const char* const name = category + pos;
+            bool found = false;
+            auto it = std::lower_bound(filters->begin(), filters->end(), std::make_pair(name, size),
+                    [&found](const FilterData &filter, const std::pair<const char*, size_t> &value) {
+                const int cmp = std::strncmp(filter.name, value.first, std::min(filter.size, value.second));
+                if (cmp == 0) {
+                    if (filter.size == value.second) {
+                        found = true;
+                    }
+                    return filter.size < value.second;
+                }
+                return cmp < 0;
+            });
+            if (!found) {
+                it = filters->insert(it, FilterData(name, size)); // Add subcategory
+            }
+            if (!category[i]) {
+                it->level = level;
+                break;
+            }
+            filters = &it->filters;
+            pos = i + 1;
+        }
+    }
+}
+
+spark::LogHandler::~LogHandler() {
+}
+
+LogLevel spark::LogHandler::categoryLevel(const char *category) const {
+    if (!category || filters_.empty()) {
+        return level_; // Default level
+    }
+    LogLevel level = level_;
+    const std::vector<FilterData> *filters = &filters_; // Root categories
+    size_t pos = 0;
+    for (size_t i = 0;; ++i) {
+        if (category[i] && category[i] != '.') { // Category separator
+            continue;
+        }
+        const size_t size = i - pos;
+        if (!size) {
+            break; // Invalid category name
+        }
+        const char* const name = category + pos;
+        bool found = false;
+        auto it = std::lower_bound(filters->begin(), filters->end(), std::make_pair(name, size),
```

Comment:
Despite looking similarly, the code in `categoryLevel()` uses const iterators while constructor deals with non-const ones. This makes refactoring that you're mentioning a bit tricky. I suggest to leave this part of the code as it is for now, as it's going to be refactored for cloud-enabled logging anyway (it's better to get rid of vectors in case of dynamically allocated loggers).